File: theraconnect_flutter/lib/models/patient.dart
class Patient {
  final String id;
  final String userId;
  final bool hasAvatar;
  final String? assignedClinicianId;
  final List<String> assignedClinicianIds;
  final String? requestedClinicianId;
  final String? clinicianRequestStatus;
  final String? dateOfBirth;
  final String? gender;
  final String? educationalAttainment;
  final String? employmentStatus;
  final String? personalIssues;
  final String? contactNo;
  final String? address;
  final String? emergencyContact;
  final String? notes;
  final String? createdAt;
  final String? updatedAt;

  const Patient({
    required this.id,
    required this.userId,
    this.hasAvatar = false,
    this.assignedClinicianId,
    this.assignedClinicianIds = const [],
    this.requestedClinicianId,
    this.clinicianRequestStatus,
    this.dateOfBirth,
    this.gender,
    this.educationalAttainment,
    this.employmentStatus,
    this.personalIssues,
    this.contactNo,
    this.address,
    this.emergencyContact,
    this.notes,
    this.createdAt,
    this.updatedAt,
  });

  factory Patient.fromJson(Map<String, dynamic> json) {
    return Patient(
      id: json['public_id'] as String,
      userId: json['user_public_id'] as String,
      hasAvatar: json['has_avatar'] as bool? ?? false,
      assignedClinicianId: json['assigned_clinician_public_id'] as String?,
      assignedClinicianIds:
          (json['assigned_clinician_public_ids'] as List<dynamic>?)
              ?.map((id) => id as String)
              .toList() ??
          const [],
      requestedClinicianId:
          json['requested_clinician_public_id'] as String?,
      clinicianRequestStatus: json['clinician_request_status'] as String?,
      dateOfBirth: json['date_of_birth'] as String?,
      gender: json['gender'] as String?,
      educationalAttainment: json['educational_attainment'] as String?,
      employmentStatus: json['employment_status'] as String?,
      personalIssues: json['personal_issues'] as String?,
      contactNo: json['contact_no'] as String?,
      address: json['address'] as String?,
      emergencyContact: json['emergency_contact'] as String?,
      notes: json['notes'] as String?,
      createdAt: json['created_at'] as String?,
      updatedAt: json['updated_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'public_id': id,
      'user_public_id': userId,
      'has_avatar': hasAvatar,
      'assigned_clinician_public_id': assignedClinicianId,
      'assigned_clinician_public_ids': assignedClinicianIds,
      'requested_clinician_public_id': requestedClinicianId,
      'clinician_request_status': clinicianRequestStatus,
      'date_of_birth': dateOfBirth,
      'gender': gender,
      'educational_attainment': educationalAttainment,
      'employment_status': employmentStatus,
      'personal_issues': personalIssues,
      'contact_no': contactNo,
      'address': address,
      'emergency_contact': emergencyContact,
      'notes': notes,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }

  static const genders = ['Male', 'Female', 'Other', 'Prefer not to say'];
  static const educationLevels = ['None', 'Elementary', 'High School', 'Vocational', 'College', 'Postgraduate'];
  static const employmentStatuses = ['Employed', 'Self-employed', 'Unemployed', 'Student', 'Retired'];
}

File: app/Models/Patient.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Patient extends Model
{
    use SoftDeletes;

    public const GENDERS = ['Male', 'Female', 'Other', 'Prefer not to say'];

    public const EDUCATION_LEVELS = ['None', 'Elementary', 'High School', 'Vocational', 'College', 'Postgraduate'];

    public const EMPLOYMENT_STATUSES = ['Employed', 'Self-employed', 'Unemployed', 'Student', 'Retired'];

    // Lifecycle of a patient's request to be assigned a clinician at sign-up.
    public const REQUEST_PENDING = 'pending';

    public const REQUEST_APPROVED = 'approved';

    public const REQUEST_DENIED = 'denied';

    protected $fillable = [
        'user_id',
        'assigned_clinician_id',
        'requested_clinician_id',
        'clinician_request_status',
        'date_of_birth',
        'gender',
        'educational_attainment',
        'employment_status',
        'personal_issues',
        'contact_no',
        'address',
        'emergency_contact',
        'notes',
    ];

    protected function casts(): array
    {
        return [
            'date_of_birth' => 'date',
            'personal_issues' => 'encrypted',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function assignedClinician(): BelongsTo
    {
        return $this->belongsTo(Clinician::class, 'assigned_clinician_id');
    }

    public function assignedClinicians(): BelongsToMany
    {
        return $this->belongsToMany(Clinician::class)->withTimestamps();
    }

    public function assignClinician(Clinician|int $clinician): void
    {
        $clinicianId = $clinician instanceof Clinician ? $clinician->id : $clinician;

        $this->assignedClinicians()->syncWithoutDetaching([$clinicianId]);

        // Compatibility for older clients and code during the pivot rollout.
        if ($this->assigned_clinician_id === null) {
            $this->update(['assigned_clinician_id' => $clinicianId]);
        }
    }

    public function isAssignedTo(Clinician|int $clinician): bool
    {
        $clinicianId = $clinician instanceof Clinician ? $clinician->id : $clinician;

        if ($this->relationLoaded('assignedClinicians')
            && $this->assignedClinicians->contains('id', $clinicianId)) {
            return true;
        }

        return $this->assigned_clinician_id === $clinicianId
            || $this->assignedClinicians()->whereKey($clinicianId)->exists();
    }

    public function scopeAssignedTo(Builder $query, Clinician|int $clinician): Builder
    {
        $clinicianId = $clinician instanceof Clinician ? $clinician->id : $clinician;

        return $query->where(function (Builder $query) use ($clinicianId) {
            $query->where('assigned_clinician_id', $clinicianId)
                ->orWhereHas('assignedClinicians', fn (Builder $q) => $q->whereKey($clinicianId));
        });
    }

    public function requestedClinician(): BelongsTo
    {
        return $this->belongsTo(Clinician::class, 'requested_clinician_id');
    }

    public function appointments(): HasMany
    {
        return $this->hasMany(Appointment::class);
    }

    public function conversations(): HasMany
    {
        return $this->hasMany(Conversation::class);
    }

    public function assignments(): HasMany
    {
        return $this->hasMany(Assignment::class);
    }

    public function submissions(): HasMany
    {
        return $this->hasMany(Submission::class);
    }

    public function clinicianNotes(): HasMany
    {
        return $this->hasMany(PatientNote::class);
    }

    public function assessments(): HasMany
    {
        return $this->hasMany(Assessment::class);
    }

    public function moodLogs(): HasMany
    {
        return $this->hasMany(MoodLog::class);
    }

    public function therapyGoals(): HasMany
    {
        return $this->hasMany(TherapyGoal::class);
    }
}
